go/vt/schemadiff/table.go

```diff
@@ -630,44 +632,37 @@ func (c *CreateTableEntity) diffKeys(alterTable *sqlparser.AlterTable,
 	}
 }
 
-// evaluateColumnReordering produces a minimal uni-directional reordering set of columns. To elaborate:
+// evaluateColumnReordering produces a minimal reordering set of columns. To elaborate:
 // The function receives two sets of columns. the two must be permutations of one another. Specifically,
 // these are the columns shared between the from&to tables.
-// The function evaluates the minimal number of steps such that:
-// - each steps reorders a column, specificly moving it backwards (never forward)
-// - the final set of steps permutate the "from" columns order in to the "to" columns order
+// The function uses longest-common-subsequence (lcs) algorithm to compute which columns should not be moved.
+// any column not in the lcs need to be reordered.
+// The function a map of column names that need to be reordered, and the index into which they are reordered.
 func evaluateColumnReordering(t1SharedColumns, t2SharedColumns []*sqlparser.ColumnDefinition) map[string]int {
 	minimalColumnReordering := map[string]int{}
-	// buf is an ever changing space. It begins with the "from" set of columns.
-	// In neach step of th ealgorithm we will remove one element from the buffer. So each
-	// step the buffer will become smaller, with less work to be done
-	buf := t1SharedColumns[:]
-	colIndexInBuf := func(name string) int {
-		for i := range buf {
-			if buf[i].Name.String() == name {
-				return i
-			}
-		}
-		return -1
+
+	t1SharedColNames := []interface{}{}
+	for _, col := range t1SharedColumns {
+		t1SharedColNames = append(t1SharedColNames, col.Name.String())
+	}
+	t2SharedColNames := []interface{}{}
+	for _, col := range t2SharedColumns {
+		t2SharedColNames = append(t2SharedColNames, col.Name.String())
+	}
+
+	lcs := golcs.New(t1SharedColNames, t2SharedColNames)
+	lcsNames := map[string]bool{}
+	for _, v := range lcs.Values() {
+		lcsNames[v.(string)] = true
 	}
-	// t2Col is our desired state. It's the "to" table's ordered list of columns
 	for i, t2Col := range t2SharedColumns {
-		// for each position in the target list of columns, we check whether we need to
-		// reorder a column.
-		// t2ColName is the desired column name at this position:
 		t2ColName := t2Col.Name.String()
-		if t2ColName == buf[0].Name.String() {
-			// no need to reorder this column. continue
-			buf = buf[1:]
-			continue
+		// see if this column is in longest common subsequence. If so, need to reorder it. If not, it must be reordered.
```

Comment:
I think the comment is wrong here, "if so, no need to reorder it"?

Reply:
good catch. Comment fixed.